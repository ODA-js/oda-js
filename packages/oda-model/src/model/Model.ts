import { Map, Record } from 'immutable';

import { IModel, IModelInit, IModelStore, IModelTransform } from '../interfaces/IModel';
import { IPackage, IPackageInit } from '../interfaces/IPackage';
import { Package } from './Package';
import { Persistent } from './Persistent';

// tslint:disable-next-line:variable-name
export const DefaultModel: IModelStore = {
  name: null,
  title: null,
  description: null,
  packages: null,
};

// tslint:disable-next-line:variable-name
export const ModelTransform: IModelTransform = {
  packages: {
    transform: (input: IPackageInit[], model: IModel) => {
      if (input) {
        return Map<string, IPackage>(input.map(p => [p.name, new Package({
          ...p,
          model,
        },
        )]) as [string, IPackage][]);
      } else {
        return null;
      }
    },
    reverse: (input?: Map<string, IPackage>) => {
      if (input) {
        return Array.from(input.values()[Symbol.iterator]()).map(f => f.toJS() as IPackageInit);
      } else {
        return null;
      }
    },
  },
};

// tslint:disable-next-line:variable-name
export const ModelStorage = Record(DefaultModel);

export class Model extends Persistent<IModelInit, IModelStore> implements IModel {
  public get modelType(): 'model' {
    return 'model';
  }
  public get name(): string {
    return this.store.get('name', null);
  }
  public get title(): string {
    return this.store.get('title', null);
  }
  public get description(): string {
    return this.store.get('description', null);
  }
  public get packages(): Map<string, IPackage> {
    return this.store.get('packages', null);
  }
  public get defaultPackage(): IPackage {
    return this.packages.get(this.defaultPackageName);
  }

  public get defaultPackageName(): string {
    return this._defaultPackageName;
  }

  private _defaultPackageName: string = 'system';

  protected transform(input: Partial<IModelInit>): IModelStore {
    const result: IModelStore = {} as any;
    if (input) {
      for (let f in input) {
        if (input.hasOwnProperty(f)) {
          if (f === 'packages') {
            result.packages = ModelTransform.packages.transform(input.packages, this);
          } else {
            result[f] = input[f];
          }
        }
      }
    }
    return result;
  }

  protected reverse(input: Record<IModelStore> & Readonly<IModelStore>): IModelInit {
    const result: IModelInit = {} as any;
    if (input) {
      const core = input.toJS();
      for (let f in core) {
        if (core.hasOwnProperty(f)) {
          if (f === 'packages') {
            result.packages = ModelTransform.packages.reverse(input.packages);
          } else {
            result[f] = core[f];
          }
        }
      }
    }
    return result;
  }

  constructor(init: Partial<IModelInit> = {
    defaultPackageName: 'system',
  }) {
    super();

    if (init.hasOwnProperty('defaultPackageName') && init.defaultPackageName && init.defaultPackageName !== this.defaultPackageName) {
      this._defaultPackageName = init.defaultPackageName;
    }
    this.store = new ModelStorage(this.transform(init));
    this.init = new (Record<Partial<IModelInit>>(init))();
    if (!this.packages || !this.packages.has(this.defaultPackageName)) {
      this.updateWith({
        packages: [{
          acl: 0,
          name: this.defaultPackageName,
          items: [],
          model: this,
        }],
      });
    }
  }
}
