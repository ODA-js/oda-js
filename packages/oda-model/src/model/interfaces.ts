import { EntityReference } from './entityreference';
import { Field } from './field';
import { RelationBase } from './relationbase';

export type RelationType = 'HasMany' | 'HasOne' | 'BelongsToMany' | 'BelongsTo';

export type MetaModelType =
  'model'
  |'package'
  | 'entity'
  | 'field'
  | 'relation'
  | 'ref'
  | RelationType
  ;

export interface IModelType extends IValidate {
  modelType: MetaModelType;
}

export interface IModel extends IModelType {
  name: string;
  packages: Map<string, IPackage>;
}

export interface IPackage extends IModelType {
  name: string;
  metaModel: IModel;
  entities: Map<string, IEntity>;
}

export interface IEntity extends IModelType {
  name: string;
  plural: string;
  fields: Map<string, Field>;
}

export interface IField extends IModelType {
  name: string;
  type: string;
  relation: IRelation;
}

export interface IRelation extends IModelType {
  verb: RelationType;
  ref: IEntityRef;
}

export interface IBelongsToManyRelation extends IRelation {

  belongsToMany: IEntityRef;
}

export interface IBelongsToRelation extends IRelation {
  belongsTo: IEntityRef;
}

export interface IHasOneRelation extends IRelation {
  hasOne: IEntityRef;
}

export interface IHasManyRelation extends IRelation {
  hasMany: IEntityRef;
}

export interface IEntityRef {
  backField: string;
  entity: string;
  field: string;
}

export type ModelItem = IModel | IPackage | IEntity | IField | IRelation;

export type Relation = IHasManyRelation | IHasOneRelation | IBelongsToRelation | IBelongsToRelation;

export function isModel(item: ModelItem ): item is IModel {
  return item.modelType === 'model';
}

export function isPackage(item: ModelItem ): item is IPackage {
  return item.modelType === 'package';
}

export function isEntity(item: ModelItem): item is IEntity {
  return item.modelType === 'entity';
}

export function isField(item: ModelItem): item is IField {
  return item.modelType === 'field';
}

export function isRelation(item: ModelItem): item is IRelation {
  return (
    item.modelType === 'BelongsTo'
    || item.modelType === 'BelongsToMany'
    || item.modelType === 'HasOne'
    || item.modelType === 'HasMany'
  );
}

export function IsBelongsTo(item: Relation): item is IBelongsToRelation {
  return isRelation(item) && item.modelType === 'BelongsTo';
}

export function IsBelongsToMany(item: Relation): item is IBelongsToRelation {
  return isRelation(item) && item.modelType === 'BelongsToMany';
}

export function IsHasOne(item: Relation): item is IBelongsToRelation {
  return isRelation(item) && item.modelType === 'HasOne';
}

export function IsHasMany(item: Relation): item is IBelongsToRelation {
  return isRelation(item) && item.modelType === 'HasMany';
}

export enum ValidationResultType { error = 'error', warning = 'warning', critics = 'critics', fixable = 'fixable' }

export interface IValidationResult {
  model?: string;
  package?: string;
  entity?: string;
  field?: string;
  result: ValidationResultType;
  message?: string;
}

export interface IValidator {
  check(item: IValidate): IValidationResult[];
}

export interface IValidate {
  validate(validator: IValidator): IValidationResult[];
}

export interface FieldInput extends FieldBaseInput {
  type?: string;
  identity?: boolean | string | string[];
  indexed?: boolean | string | string[];
  required?: boolean;
  arguments?: [FieldArgs];
  relation?: ({ hasMany: string }
    | { hasOne: string }
    | { belongsTo: string }
    | { belongsToMany: string, using: string }) & {
    entity: string;
    field: string;
  };
}

export interface FieldStorage extends FieldBaseStorage {
  type: string;
  arguments?: [FieldArgs];
  type_: string;
  idKey: EntityReference;
  relation: RelationBase;
}

export interface BelongsToInput extends RelationBaseInput {
  belongsTo?: string;
}

export interface BelongsToStorage extends RelationBaseStorage {
  belongsTo: EntityReference;
  belongsTo_: string;
}

export interface BelongsToManyInput extends RelationBaseInput {
  belongsToMany: string;
  using: string;
}

export interface BelongsToManyStorage extends RelationBaseStorage {
  belongsToMany: EntityReference;
  belongsToMany_?: string;
  using: EntityReference;
  using_?: string;
}

export interface EntityInput extends ModelBaseInput {
  plural?: string;
  fields?: FieldInput[] | {
    [fName: string]: FieldInput;
  };
}

export interface EntityJSON extends ModelBaseInput {
  fields?: FieldInput[];
}

export interface EntityStorage extends ModelBaseStorage {
  fields: Map<string, Field>;
  relations: Set<string>;
  identity: Set<string>;
  required: Set<string>;
  indexed: Set<string>;
}

export interface EntityReferenceInput {
  backField: string;
  field: string;
  entity: string;
}

export interface FieldArgs {
  name: string;
  type: string;
  required: boolean;
  defaultValue: string;
}

export interface FieldBaseInput extends ModelBaseInput {
  args?: [FieldArgs];
  derived?: boolean;
  persistent?: boolean;
  entity?: string;
}

export interface FieldBaseStorage extends ModelBaseStorage {
  args?: [FieldArgs];
  args_?: [FieldArgs];
  entity: string;
  entity_: string;
}

export interface HasManyInput extends RelationBaseInput {
  hasMany: string;
}

export interface HasManyStorage extends RelationBaseStorage {
  hasMany: EntityReference;
  hasMany_: string;
}

export interface HasOneInput extends RelationBaseInput {
  hasOne: string;
}

export interface HasOneStorage extends RelationBaseStorage {
  hasOne: EntityReference;
  hasOne_: string;
}

export interface MetadataInput {
  metadata?: { [key: string]: any };
}

export interface ModelBaseInput extends MetadataInput {
  name: string;
  title?: string;
  description?: string;
}

export interface ModelBaseStorage {
  name: string;
  title: string;
  description: string;
  name_: string;
  title_: string;
  description_: string;
}

export interface ModelPackageInput extends ModelBaseInput {
  name: string;
  title?: string;
  description?: string;
  abstract?: boolean;
  entities: string[];
  mutations: any[];
}

export interface ModelPackageStore {
  name: string;
  title?: string;
  description?: string;
  entities: string[];
  mutations: any[];
}

export interface MetaModelStore {
  entities: EntityInput[];
  packages: ModelPackageStore[];
  mutations: MutationInput[];
  name: string;
  title?: string;
  description?: string;
}


export interface RelationBaseInput {
  /**
   * нужно в случае когда мы будем показывать атрибут связи, и ассоциацию отедельно???
   * больше не зачем
   */
  metadata?: { [key: string]: any };
  name?: string;
  entity: string;
  field: string;
  fields?: FieldInput[] | {
    [field: string]: FieldInput,
  };
  opposite?: string;
}

export interface RelationBaseJSON {
  /**
   * нужно в случае когда мы будем показывать атрибут связи, и ассоциацию отедельно???
   * больше не зачем
   */
  metadata?: { [key: string]: any };
  name?: string;
  fields?: FieldInput[];
  opposite?: string;
}

export interface RelationFields {
  description: string;
  name: string;
  type: string;
  required: boolean;
  indexed: boolean | string | string[];
  identity?: boolean | string | string[];
  derived: boolean;
  persistent: boolean;
  args: FieldArgs[];
}

export interface RelationBaseStorage {
  name?: string;
  name_?: string;
  entity: string;
  entity_: string;
  field: string;
  field_: string;
  fields: Map<string, Field>;
  opposite: string;
}

export interface MutationInput extends ModelBaseInput {
  args: FieldArgs[];
  payload: FieldArgs[];
}

export interface MutationStorage extends ModelBaseStorage {
  args: FieldArgs[];
  args_: FieldArgs[];
  payload: FieldArgs[];
  payload_: FieldArgs[];
}

export interface ModelHook {
  name: string;
  entities?: {
    [eName: string]: EntityInput,
  };
  mutations?: {
    [mName: string]: MutationInput;
  };
}
