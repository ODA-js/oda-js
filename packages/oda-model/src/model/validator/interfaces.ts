import { BelongsTo } from '../belongsto';
import { BelongsToMany } from '../belongstomany';
import { HasMany } from '../hasmany';
import { HasOne } from '../hasone';
import { IEntity, IField, IModel, IPackage, IRelation } from '../interfaces';

export interface IModelContext {
  model: IModel;
}

export interface IPackageContext {
  model: IModel;
  package: IPackage;
}

export interface IEntityContext {
  model: IModel;
  package: IPackage;
  entity: IEntity;
}

export interface IFieldContext {
  model: IModel;
  package: IPackage;
  entity: IEntity;
  field: IField;
}

export type Relation =
  HasMany
  | HasOne
  | BelongsToMany
  | BelongsTo;

export interface IRelationContext {
  model: IModel;
  package: IPackage;
  entity: IEntity;
  field: IField;
  relation: IRelation;
}

export type ValidationContext =
  IModelContext | IPackageContext | IEntityContext | IFieldContext | IRelationContext;
