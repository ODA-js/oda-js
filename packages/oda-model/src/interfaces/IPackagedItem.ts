import { IPackage } from './IPackage';
import { IContext } from '../contexts/IContext';
import { IPackageContext } from '../contexts/IPackageContext';

export type PackagedItemInit = IPackagedItemInit | string;

export interface IPackagedItemInit {
  name: string;
  context: IPackageContext;
}

export interface IPackagedItemStore {
  name: string;
  context: IPackageContext;
}

export interface IPackagedItem {
  readonly name: string;
  readonly context: IPackageContext;
  toJS(): object;
}
