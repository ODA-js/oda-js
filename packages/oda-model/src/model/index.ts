import { Entity } from './entity';
import { EntityBase } from './entitybase';
import { Mixin } from './mixin';
import { Union } from './union';
import { EntityReference } from './entityreference';
import { Field } from './field';
import { FieldBase } from './fieldbase';
import { HasOne } from './hasone';
import { HasMany } from './hasmany';
import { BelongsTo } from './belongsto';
import { BelongsToMany } from './belongstomany';
import { MetaModel } from './metamodel';
import { ModelBase } from './modelbase';
import { Mutation } from './mutation';
import { DEFAULT_ID_FIELD } from './definitions';
import { ModelPackage } from './modelpackage';
import { RelationBase } from './relationbase';
import {
  FieldArgs, MetaModelStore,
  ModelHook, IValidationResult, ValidationResultType,
} from './interfaces';
import { Metadata } from './metadata';
import Validator from './validator/index';

export {
  ModelHook,
  Metadata,
  FieldArgs,
  MetaModelStore,
  Entity,
  Field,
  HasOne,
  HasMany,
  BelongsTo,
  BelongsToMany,
  ModelPackage,
  MetaModel,
  DEFAULT_ID_FIELD,
  Mutation,
  RelationBase,
  FieldBase,
  EntityReference,
  ModelBase,
  IValidationResult,
  ValidationResultType,
  Validator,
};
