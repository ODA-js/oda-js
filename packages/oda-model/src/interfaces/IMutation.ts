import { Map, Set } from 'immutable';

import { ArrayToMap, ArrayToSet, MapType } from '../model/utils';
import { IFieldArgInit, IFieldArg, FieldArgsInput } from './IFieldArg';
import { IModelType, INamedItem } from './IModelType';
import { IPackage } from './IPackage';
import { IPackagedItem, IPackagedItemInit, IPackagedItemStore } from './IPackagedItem';
import { FieldArgsTransformType } from './types';

export interface IMutationACL {
  execute: string[];
}

export interface IMutationMetaData {
  acl?: IMutationACL;
}

export interface IMutationACLStore {
  execute: Set<string>;
}

export interface IMutationMetaDataStore {
  acl?: IMutationACLStore;
}

export interface IMutationStore extends IMutationMetaDataStore, INamedItem, IPackagedItemStore {
  args: Map<string, IFieldArg>;
  payload: Map<string, IFieldArg>;
}

export interface IMutationInit extends Partial<IMutationMetaData>, INamedItem, IPackagedItemInit {
  args: FieldArgsInput;
  payload: FieldArgsInput;
  package?: IPackage;
}

export interface IMutationACLTransform {
  execute: ArrayToSet<string>;
}

export interface IMutationTransform {
  args: FieldArgsTransformType;
  payload: FieldArgsTransformType;
  acl: IMutationACLTransform;
}

export interface IMutation extends IModelType, IPackagedItem {
  readonly modelType: 'mutation';
  readonly args: Map<string, IFieldArgInit>;
  readonly payload: Map<string, IFieldArgInit>;
  readonly acl?: IMutationACLStore;
}

