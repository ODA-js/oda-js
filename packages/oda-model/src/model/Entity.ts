import { Map, Record, Set } from 'immutable';

import { IEntity, IEntityInit, IEntityStore, IEntityTransform, IEntityACL} from '../interfaces/IEntity';
import { IField } from '../interfaces/IField';
import { IPackage } from '../interfaces/IPackage';
import { Persistent } from './Persistent';
import { TransformField } from './utils';
import { IPackageContext } from '../contexts/IPackageContext';
import { IEntityContext } from '../contexts/IEntityContext';

// tslint:disable-next-line:variable-name
export const DefaultEntity: IEntityStore = {
  name: null,
  title: null,
  description: null,
  acl: null,
  plural: null,
  singular: null,
  fields: null,
  storage: null,
  indexed: null,
  relations: null,
  required: null,
};

// tslint:disable-next-line:variable-name
export const EntityTransform: IEntityTransform = {
  fields: TransformField(),
};

// tslint:disable-next-line:variable-name
export const EntityStorage = Record(DefaultEntity);

export class Entity extends Persistent<IEntityInit, IEntityStore, IPackageContext> implements IEntity {
  public get modelType(): 'entity' {
    return 'entity';
  }
  public get name(): string {
    return this.store.get('name', null);
  }
  public get title(): string {
    return this.store.get('title', null);
  }
  public get acl() {
    return this.store.get('acl', null);
  }
  public get description(): string {
    return this.store.get('description', null);
  }
  public get singular(): string {
    return this.store.get('singular', null);
  }
  public get plural(): string {
    return this.store.get('plural', null);
  }
  public get fields(): Map<string, IField> {
    return this.store.get('fields', null);
  }
  public get relations(): Set<string> {
    return this.store.get('relations', null);
  }
  public get required(): Set<string> {
    return this.store.get('required', null);
  }
  public get indexed(): Set<string> {
    return this.store.get('indexed', null);
  }
  protected transform(input: Partial<IEntityInit>): IEntityStore {
    const result: IEntityStore = {} as any;
    if (input) {
      for (let f in input) {
        if (input.hasOwnProperty(f)) {
          if (f === 'fields') {
            result.fields = EntityTransform.fields.transform(input.fields, this);
          } else {
            result[f] = input[f];
          }
        }
      }
    }
    return result;
  }

  protected reverse(input: Record<IEntityStore> & Readonly<IEntityStore>): IEntityInit {
    const result: IEntityInit = {} as any;
    if (input) {
      const core = input.toJS();
      for (let f in core) {
        if (core.hasOwnProperty(f)) {
          if (f === 'fields') {
            result.fields = EntityTransform.fields.reverse(input.fields);
          } else {
            result[f] = core[f];
          }
        }
      }
    }
    return result;
  }

  constructor(init?: Partial<IEntityInit>, context?: IPackageContext) {
    super();
    if (!context && init && init.fields) {
      throw new Error('context must be provided');
    }
    if (context) {
      this.attach(context);
    }
    this.store = new EntityStorage(this.transform(init));
  }
}
