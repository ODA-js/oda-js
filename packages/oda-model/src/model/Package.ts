import { Map, Record } from 'immutable';

import { IModelContext } from '../contexts/IModelContext';
import { IPackage, IPackageInit, IPackageStore, IPackageTransform } from '../interfaces/IPackage';
import { IPackagedItem, IPackagedItemInit, PackagedItemInit } from '../interfaces/IPackagedItem';
import { Persistent } from './Persistent';

// tslint:disable-next-line:variable-name
export const DefaultPackage: IPackageStore = {
  name: null,
  title: null,
  description: null,
  acl: null,
  abstract: null,
  items: null,
};

// tslint:disable-next-line:variable-name
export const PackageTransform: IPackageTransform = {
  items: {
    transform: (input: PackagedItemInit[], pkg: Package) => {
      if (input) {
        return Map<string, IPackagedItem>(input.map(p => {
          if (typeof p === 'string') {
            if (pkg.context.model.defaultPackage.items.has(p)) {
              return [p, {
                ...pkg.context.model.defaultPackage.items.get(p),
                package: pkg,
              }];
            } else {
              throw Error('item does not exists');
            }
          } else {
            return [p.name, {
              ...p,
              package: pkg,
            }];
          }
        },
        ) as [string, IPackagedItem][]);
      } else {
        return null;
      }
    },
    reverse: (input?: Map<string, IPackagedItem>) => {
      if (input) {
        return Array.from(input.values()[Symbol.iterator]()).map(i => i.toJS() as IPackagedItemInit);
      } else {
        return null;
      }
    },
  },
};

// tslint:disable-next-line:variable-name
const PackageStorage = Record(DefaultPackage);

export class Package extends Persistent<IPackageInit, IPackageStore, IModelContext> implements IPackage {
  public get modelType(): 'package' {
    return 'package';
  }
  public get name(): string {
    return this.store.get('name', null);
  }
  public get description(): string {
    return this.store.get('description', null);
  }
  public get abstract(): boolean {
    return this.store.get('abstract', null);
  }
  public get title(): string {
    return this.store.get('title', null);
  }
  public get acl(): number {
    return this.store.get('acl', null);
  }
  public get items(): Map<string, IPackagedItem> {
    return this.store.get('items', null);
  }

  protected transform(input: Partial<IPackageInit>): IPackageStore {
    const result: IPackageStore = {} as any;
    if (input) {
      for (let f in input) {
        if (input.hasOwnProperty(f)) {
          if (f === 'items') {
            result.items = PackageTransform.items.transform(input.items, this);
          } else {
            result[f] = input[f];
          }
        }
      }
    }
    return result;
  }

  protected reverse(input: Record<IPackageStore> & Readonly<IPackageStore>): IPackageInit {
    const result: IPackageInit = {} as any;
    if (input) {
      const core = input.toJS();
      for (let f in core) {
        if (core.hasOwnProperty(f)) {
          if (f === 'items') {
            result.items = PackageTransform.items.reverse(input.items);
          } else {
            result[f] = core[f];
          }
        }
      }
    }
    return result;
  }

  constructor(init: Partial<IPackageInit>, context: IModelContext) {
    super();
    if (!context && init.items) {
      throw new Error('context must be provided');
    }
    this.attach(context);
    this.store = new PackageStorage(this.transform(init));
    this.init = new (Record<Partial<IPackageInit>>(init))();
  }
}
