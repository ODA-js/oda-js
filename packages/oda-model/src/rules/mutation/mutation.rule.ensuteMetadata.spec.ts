import { IPackagedItemInit } from '../../interfaces/IPackagedItem';
import { Entity } from '../../model/Entity';
import { Model } from '../../model/Model';
import { Package } from '../../model/Package';
import Rule from './ensureMetadata';
import { IMutationInit } from '../../interfaces/IMutation';

describe('rule', () => {
  let rule: Rule;

  const models: {
    model: Model,
    package: Package,
  } = {} as any;

  beforeEach(() => {
    rule = new Rule();
    expect(() => models.model = new Model({
      name: 'TodoItems',
    })).not.toThrow();
    models.package = models.model.packages.get('system') as Package;
  });

  it('fix acl', () => {
    models.package.updateWith({
      items: [{
        name: 'mutation',
        args: {
          input: {},
        },
        payload: {
          result: {},
        },
      } as IMutationInit],
    });
    debugger;
    const mutation = models.package.items.get('mutation');
    expect(mutation).toMatchSnapshot();
    const result = rule.validate({
      model: {},
      package: {},
      mutation,
    } as any);
    expect(mutation).toMatchSnapshot();
  });

});
