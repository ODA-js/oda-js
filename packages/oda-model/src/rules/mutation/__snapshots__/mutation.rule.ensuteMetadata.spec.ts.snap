// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rule fix acl 1`] = `
Mutation {
  "_context": Immutable.Record {
    "model": null,
    "package": Package {
      "_context": Immutable.Record {
        "model": Model {
          "_defaultPackageName": "system",
          "store": Immutable.Record {
            "name": "TodoItems",
            "title": null,
            "description": null,
            "packages": Immutable.Map {
              "system": [Circular],
            },
          },
        },
      },
      "store": Immutable.Record {
        "name": "system",
        "title": null,
        "description": null,
        "acl": 0,
        "abstract": null,
        "items": Immutable.Map {
          "mutation": [Circular],
        },
      },
    },
  },
  "store": Immutable.Record {
    "name": "mutation",
    "title": null,
    "description": null,
    "args": Immutable.Map {
      "input": FieldArg {
        "_context": Immutable.Record {
          "model": null,
          "package": null,
          "mutation": [Circular],
        },
        "store": Immutable.Record {
          "name": "input",
          "title": null,
          "description": null,
          "required": null,
          "type": null,
          "defaultValue": null,
        },
      },
    },
    "payload": Immutable.Map {
      "result": FieldArg {
        "_context": Immutable.Record {
          "model": null,
          "package": null,
          "mutation": [Circular],
        },
        "store": Immutable.Record {
          "name": "result",
          "title": null,
          "description": null,
          "required": null,
          "type": null,
          "defaultValue": null,
        },
      },
    },
    "acl": null,
  },
}
`;

exports[`rule fix acl 2`] = `
Mutation {
  "_context": Immutable.Record {
    "model": null,
    "package": Package {
      "_context": Immutable.Record {
        "model": Model {
          "_defaultPackageName": "system",
          "store": Immutable.Record {
            "name": "TodoItems",
            "title": null,
            "description": null,
            "packages": Immutable.Map {
              "system": [Circular],
            },
          },
        },
      },
      "store": Immutable.Record {
        "name": "system",
        "title": null,
        "description": null,
        "acl": 0,
        "abstract": null,
        "items": Immutable.Map {
          "mutation": [Circular],
        },
      },
    },
  },
  "store": Immutable.Record {
    "name": "mutation",
    "title": null,
    "description": null,
    "args": Immutable.Map {
      "input": FieldArg {
        "_context": Immutable.Record {
          "model": null,
          "package": null,
          "mutation": [Circular],
        },
        "store": Immutable.Record {
          "name": "input",
          "title": null,
          "description": null,
          "required": null,
          "type": null,
          "defaultValue": null,
        },
      },
    },
    "payload": Immutable.Map {
      "result": FieldArg {
        "_context": Immutable.Record {
          "model": null,
          "package": null,
          "mutation": [Circular],
        },
        "store": Immutable.Record {
          "name": "result",
          "title": null,
          "description": null,
          "required": null,
          "type": null,
          "defaultValue": null,
        },
      },
    },
    "acl": Object {
      "execute": Immutable.Set [],
    },
  },
}
`;
